treq/testing.py

```diff
@@ -205,3 +207,212 @@ def __init__(self, resource):
+        absoluteURI = str(request.URLPath().click(request.path))
+
+        status_code, headers, body = self._get_response_for(
+            request.method, absoluteURI, params, headers,
+            request.content.read())
+
+        request.setResponseCode(status_code)
+        for k, v in headers.items():
+            request.setHeader(k, v)
+
+        return body
+
+
+class HasHeaders(object):
+    """
+    Since Twisted adds headers to a request, such as the host and the content
+    length, it's necessary to test whether request headers CONTAIN the expected
+    headers (the ones that are not automatically added by Twisted).
+
+    This wraps a set of headers, and can be used in an equality test against
+    a superset if the provided headers.
+    """
+    def __init__(self, headers):
+        self._headers = dict([(k.lower(), v) for k, v in headers.items()])
+
+    def __repr__(self):
+        return "HasHeaders({0})".format(repr(self._headers))
+
+    def __eq__(self, other_headers):
+        compare_to = dict([(k.lower(), v) for k, v in other_headers.items()])
+
+        return (set(self._headers.keys()).issubset(set(compare_to.keys())) and
+                all([set(v).issubset(set(compare_to[k]))
+                     for k, v in self._headers.items()]))
+
+    def __ne__(self, other_headers):
+        return not self.__eq__(other_headers)
+
+
+class _any(object):
+    """
+    An object is equivalent to any other object.
+    """
+    def __repr__(self):
+        return "ANYTHING"
+
+    def __eq__(self, _):
+        return True
+
+    def __ne__(self, _):
+        return False
+
+ANY = _any()
+
+
+class SequenceStringStubs(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
+    responses, or the request's paramters do not match the next item's expected
+    request paramters, raises :obj:`AssertionError`.
+
+    :ivar list sequence: The sequence of expected request arguments mapped to
+        stubbed responses
+    :ivar testcase: A :obj:`TestCase` that can be used to report failures in
+        matching requests against expected requests.  Raising exceptions does
+        not work, because :obj:`Resource.render` will just convert that into
+        a 500 response.
+    """
+    def __init__(self, sequence, testcase):
+        self._sequence = sequence
+        self._testcase = testcase
+
+    def consumed(self):
+        """
+        :return: `bool` representing whether the entire sequence has been
+            consumed.  This is useful in tests to assert that the expected
+            requests have all been made.
+        """
+        return len(self._sequence) == 0
+
+    @contextmanager
+    def consume(self):
+        """
+        Usage::
+
+            stub_treq = StubTreq(StringStubbingResource(
+                sequence_stubs.get_response_for))
```

Comment:
I think the case of us needing to add more  methods here doesn't have a too-horrible backwards compatibility story. StringStubbingResource later wants to call more methods, it can do `getattr(o, whatever, None)` checks for the new methods that it wants to call, which it would need to do in any case. Then, if desired, `__call__` can be deprecated.

However, I think more likely is if new functionality comes about, new top-level classes/functions will be created to provide it (like, other types of test-utility IResource providers), instead of StringStubbingResource (and its response-providing object) being expanded.